File: packages/protocol/src/revocation.test.ts
import { describe, expect, it } from 'vitest';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';
import { bytesToHex, hexToBytes } from './bytes.js';
import { createDeviceIdentityFromSeed } from './identity.js';
import {
  buildRevocationChallenge,
  signRevocation,
  validateRevocationRecord,
  verifyRevocation,
  type RevocationRecord,
} from './revocation.js';

interface RevocationVector {
  name: string;
  revoker_seed_hex: string;
  revoker_device_id: string;
  revoker_pub_key_hex: string;
  revoked_device_id: string;
  seq: number;
  timestamp: string;
  challenge_hex: string;
  signature_hex: string;
  valid: boolean;
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const vectorPath = path.resolve(__dirname, '../../wire/testdata/revocation-vectors.json');

describe('Revocation Records & Cross-Language Vectors', () => {
  const vectors: RevocationVector[] = JSON.parse(fs.readFileSync(vectorPath, 'utf8'));

  for (const v of vectors) {
    it(`Vector: ${v.name}`, async () => {
      const pubKey = hexToBytes(v.revoker_pub_key_hex);
      const rec: RevocationRecord = {
        revoker_device_id: v.revoker_device_id,
        revoked_device_id: v.revoked_device_id,
        seq: v.seq,
        timestamp: v.timestamp,
        signature: v.signature_hex,
      };

      const challenge = buildRevocationChallenge(
        v.revoker_device_id,
        v.revoked_device_id,
        v.seq,
        v.timestamp,
      );
      expect(bytesToHex(challenge)).toBe(v.challenge_hex);

      const now = new Date(v.timestamp);
      const ok = await verifyRevocation(rec, pubKey, 5 * 60 * 1000, now);
      expect(ok).toBe(v.valid);
    });
  }

  it('signs and verifies valid revocation records end-to-end', async () => {
    const seed = new Uint8Array(32).fill(7);
    const id = await createDeviceIdentityFromSeed(seed);
    const targetRevoked = 'sb-dev-1111111111111111111111111111111111111111111111111111111111111111';

    const now = new Date('2026-08-30T14:00:00Z');
    const rec = await signRevocation(id, targetRevoked, 1, now);

    expect(rec.revoker_device_id).toBe(id.deviceId);
    expect(rec.revoked_device_id).toBe(targetRevoked);
    expect(rec.seq).toBe(1);
    expect(rec.timestamp).toBe('2026-08-30T14:00:00.000Z');

    validateRevocationRecord(rec);

    const valid = await verifyRevocation(rec, id.publicKey, 5 * 60 * 1000, now);
    expect(valid).toBe(true);

    // Reject self-revocation
    await expect(signRevocation(id, id.deviceId, 1, now)).rejects.toThrow(
      'cannot revoke self in mesh sync',
    );

    // Reject tampered signature
    const tamperedSig: RevocationRecord = {
      ...rec,
      signature: bytesToHex(new Uint8Array(64).fill(0xee)),
    };
    expect(await verifyRevocation(tamperedSig, id.publicKey, 5 * 60 * 1000, now)).toBe(false);

    // Reject tampered payload
    const tamperedPayload: RevocationRecord = {
      ...rec,
      revoked_device_id: 'sb-dev-2222222222222222222222222222222222222222222222222222222222222222',
    };
    expect(await verifyRevocation(tamperedPayload, id.publicKey, 5 * 60 * 1000, now)).toBe(false);
  });
});

File: packages/wire/revocation_test.go
package wire

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

type RevocationVector struct {
	Name            string `json:"name"`
	RevokerSeedHex  string `json:"revoker_seed_hex"`
	RevokerDeviceID string `json:"revoker_device_id"`
	RevokerPubKey   string `json:"revoker_pub_key_hex"`
	RevokedDeviceID string `json:"revoked_device_id"`
	Seq             uint64 `json:"seq"`
	Timestamp       string `json:"timestamp"`
	ChallengeHex    string `json:"challenge_hex"`
	SignatureHex    string `json:"signature_hex"`
	Valid           bool   `json:"valid"`
}

func TestRevocationVectors_CrossLanguage(t *testing.T) {
	vectorPath := filepath.Join("testdata", "revocation-vectors.json")
	data, err := os.ReadFile(vectorPath)
	if err != nil {
		t.Fatalf("failed to read test vectors: %v", err)
	}

	var vectors []RevocationVector
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("failed to unmarshal test vectors: %v", err)
	}

	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			pubBytes, err := hex.DecodeString(v.RevokerPubKey)
			if err != nil {
				t.Fatalf("invalid revoker pub key hex: %v", err)
			}
			pubKey := ed25519.PublicKey(pubBytes)

			rec := &RevocationRecord{
				RevokerDeviceID: v.RevokerDeviceID,
				RevokedDeviceID: v.RevokedDeviceID,
				Seq:             v.Seq,
				Timestamp:       v.Timestamp,
				Signature:       v.SignatureHex,
			}

			// Validate challenge construction
			challenge := BuildRevocationChallenge(v.RevokerDeviceID, v.RevokedDeviceID, v.Seq, v.Timestamp)
			if hex.EncodeToString(challenge) != v.ChallengeHex {
				t.Errorf("challenge mismatch:\ngot  %s\nwant %s", hex.EncodeToString(challenge), v.ChallengeHex)
			}

			now, _ := time.Parse(time.RFC3339, v.Timestamp)
			verifyErr := VerifyRevocation(rec, pubKey, 5*time.Minute, now)

			if v.Valid && verifyErr != nil {
				t.Errorf("expected valid vector to pass verification, got error: %v", verifyErr)
			}
			if !v.Valid && verifyErr == nil {
				t.Errorf("expected invalid vector to fail verification, but it passed")
			}
		})
	}
}

func TestRevocationRecord_ValidationAndErrors(t *testing.T) {
	seed := sha256.Sum256([]byte("test-revoker-seed-12345"))
	priv := ed25519.NewKeyFromSeed(seed[:])
	id, err := NewDeviceIdentity(priv.Public().(ed25519.PublicKey), priv)
	if err != nil {
		t.Fatal(err)
	}

	revokedID := "sb-dev-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	now := time.Date(2026, 8, 30, 12, 0, 0, 0, time.UTC)

	// Happy path
	rec, err := SignRevocation(id, revokedID, 1, now)
	if err != nil {
		t.Fatalf("SignRevocation failed: %v", err)
	}

	if err := rec.Validate(); err != nil {
		t.Fatalf("Validate failed on valid record: %v", err)
	}

	if err := VerifyRevocation(rec, id.PublicKey, 5*time.Minute, now); err != nil {
		t.Fatalf("VerifyRevocation failed: %v", err)
	}

	// Self-revocation attempt rejected
	_, err = SignRevocation(id, id.DeviceID, 1, now)
	if err == nil {
		t.Fatal("expected SignRevocation to reject self-revocation")
	}

	// Seq = 0 rejected
	_, err = SignRevocation(id, revokedID, 0, now)
	if err == nil {
		t.Fatal("expected SignRevocation to reject seq = 0")
	}

	// Tampered signature rejected
	tampered := *rec
	tampered.Signature = hex.EncodeToString(make([]byte, 64))
	if err := VerifyRevocation(&tampered, id.PublicKey, 5*time.Minute, now); err == nil {
		t.Fatal("expected tampered signature to fail verification")
	}

	// Tampered revoked ID rejected
	tamperedID := *rec
	tamperedID.RevokedDeviceID = "sb-dev-bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
	if err := VerifyRevocation(&tamperedID, id.PublicKey, 5*time.Minute, now); err == nil {
		t.Fatal("expected tampered revoked ID to fail verification")
	}

	// Future timestamp beyond skew rejected
	futureTime := now.Add(10 * time.Minute)
	recFuture, err := SignRevocation(id, revokedID, 1, futureTime)
	if err != nil {
		t.Fatal(err)
	}
	if err := VerifyRevocation(recFuture, id.PublicKey, 5*time.Minute, now); err == nil {
		t.Fatal("expected future timestamp beyond skew to fail verification")
	}
}
